### PR TITLE
Refactor index-as-string groupby tests and fix spurious warning (Bug 17383)

### LINKED ISSUE #17383: Groupby with matching column and index name emits spurious warning
xref https://github.com/pandas-dev/pandas/pull/14432

Which of these should raise `FutureWarning`s? I think the idea was to use `pd.Grouper(name)` to disambiguate? In that case `In [21]` should not raise a warning?

```python
In [17]: df = pd.DataFrame({"A": [1] * 5 + [2] * 5, "B": ['a', 'b'] * 5, 'C': range(10)}, index=pd.Index(range(10), name='A'))

In [19]: _ = df.groupby('A').mean()
/Users/taugspurger/.virtualenvs/pandas-dev/bin/ipython:1: FutureWarning: 'A' is both a column name and an index level.
Defaulting to column but this will raise an ambiguity error in a future version
  #!/Users/taugspurger/Envs/pandas-dev/bin/python3.6

In [20]: _ = df.groupby(['A']).mean()
/Users/taugspurger/.virtualenvs/pandas-dev/bin/ipython:1: FutureWarning: 'A' is both a column name and an index level.
Defaulting to column but this will raise an ambiguity error in a future version
  #!/Users/taugspurger/Envs/pandas-dev/bin/python3.6

In [21]: _ = df.groupby(pd.Grouper('A')).mean()
/Users/taugspurger/Envs/pandas-dev/lib/python3.6/site-packages/pandas/pandas/core/groupby.py:1699: FutureWarning: 'A' is both a column name and an index level.
Defaulting to column but this will raise an ambiguity error in a future version
  return klass(obj, by, **kwds)

In [22]: _ = df.groupby([pd.Grouper('A')]).mean()
```

cc @jmmease

### PR DESCRIPTION
**Pull request by jonmmease**
- [x] closes #17383 
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

Test case refactoring:
  - Moved the existing  index-as-string test cases out of `test_groupby.py` and into a new `test_index_as_string.py` file
  - Extracted test data generation functions and parameterized existing test cases to clean them up and shorten them.
  - Added a new parameterized test case on a `Series`
  - Updated `test_grouper_column_index_level_precedence` to reproduce false warning problem as described in #17383 
  - Updated `test_grouper_column_index_level_precedence` to verify when warnings should **NOT** be raised (Results in a test failure due to #17383 without this fix)

### CODE REVIEW

File: pandas/tests/groupby/test_groupby.py [hunk 1]
```diff
@@ -253,158 +253,6 @@ def test_grouper_column_and_index(self):
...
-                                      'inner': [1, 1, 1, 1, 1, 1]},
-                                     index=idx)
-
-        df_single_both = df_multi_both.reset_index('outer')
-
-        # Group MultiIndex by single key
-        with tm.assert_produces_warning(FutureWarning, check_stacklevel=False):
-            result = df_multi_both.groupby('inner').mean()
-
-        expected = df_multi_both.groupby([pd.Grouper(key='inner')]).mean()
```
Comment (jmmease) [inline pandas/tests/groupby/test_groupby.py]: The `pd.Grouper` object in this `expected` expression shouldn't have been wrapped in a list. If it had not been, the spurious warning would have been raised in this test. This is corrected in the new test below.

File: pandas/tests/groupby/test_index_as_string.py [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+        # Add 'inner' column to frame
+        # (frame already has an 'inner' index)
+        frame['inner'] = [1, 1, 1, 1, 1, 1]
+
+        # Group by single key
+        with tm.assert_produces_warning(FutureWarning, check_stacklevel=False):
+            result = frame.groupby('inner').mean()
+
+        with tm.assert_produces_warning(False):
+            expected = frame.groupby(pd.Grouper(key='inner')).mean()
```
Comment (jmmease) [inline pandas/tests/groupby/test_index_as_string.py]: Note that the `pd.Grouper` object is no longer wrapped in a list and that we now assert that no warning is raised. This is the test case that would have failed without the fix in this PR.

File: pandas/core/groupby.py [hunk 3]
```diff
@@ -2704,7 +2704,7 @@ def _get_grouper(obj, key=None, axis=0, level=None, sort=True,
 
     # a passed-in Grouper, directly convert
     if isinstance(key, Grouper):
-        binner, grouper, obj = key._get_grouper(obj)
+        binner, grouper, obj = key._get_grouper(obj, validate=False)
```
Comment (jreback) [inline pandas/core/groupby.py]: note that I had to add this flag to 'fix' this warning issue elsewhere, I don't really like it, but would require more refactoring to make this cleaner.

File: pandas/tests/groupby/test_index_as_string.py [hunk 4]
```diff
@@ -0,0 +1,142 @@
+import pytest
+import pandas as pd
+import numpy as np
+
+from pandas.util.testing import assert_frame_equal, assert_series_equal
+import pandas.util.testing as tm
+
+
+def build_df_multi():
+    idx = pd.MultiIndex.from_tuples([('a', 1), ('a', 2), ('a', 3),
```
Comment (jreback) [inline pandas/tests/groupby/test_index_as_string.py]: these should just be fixtures

File: pandas/tests/groupby/test_index_as_string.py [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+    df_single = build_df_multi().reset_index('outer')
+    return df_single
+
+
+def build_test_series():
+    series_multi = build_df_multi().set_index('B', append=True)['A']
+    return series_multi
+
+
+class TestGroupByIndexAsString(object):
```
Comment (jreback) [inline pandas/tests/groupby/test_index_as_string.py]: no real need to make this a class, that is really a leftover from nose, just make these functions (of course a class is good for grouping generally)

File: pandas/tests/groupby/test_index_as_string.py [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+
+        # Group by single key
+        with tm.assert_produces_warning(FutureWarning, check_stacklevel=False):
+            result = frame.groupby('inner').mean()
+
+        with tm.assert_produces_warning(False):
+            expected = frame.groupby(pd.Grouper(key='inner')).mean()
+
+        assert_frame_equal(result, expected)
+
```
Comment (jreback) [inline pandas/tests/groupby/test_index_as_string.py]: happy to have even another level of parameterization to make these shorter here (if that's possible)

**Review by jreback — COMMENTED** [review]
some comments

**Comment by jmmease** [comment]
Thanks for the feedback @jreback.  I think I've made all the changes you requested and I learned some things about pytest along the way.

**Comment by jreback** [comment]
thank @jmmease nice patch! keep em coming!

**Pull request merged**